### DISCUSSION TITLE
How to generate .d.ts files when using multiple targets

### DISCUSSION THREAD

**Discussion by captbaritone** [parcel-bundler/parcel#8815] [Q&A]
For my library I'm generating multiple targets based on different entrypoints. I'd like to include type information for each target in my published package. Is it possible to have Parcel generate `.d.ts` files for each target/bundle?

**Comment by devongovett** [comment]
The way types are generated is via a [named pipeline](https://parceljs.org/features/plugins/#named-pipelines) called "types" which is automatically inferred from the types field in package.json. You can define a named pipeline for your own targets as well and use the [same configuration](https://github.com/parcel-bundler/parcel/blob/v2/packages/configs/default/index.json#L4). So if you have a target named "frontend-types", create a pipeline with the same name.

In a .parcelrc:

```json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "frontend-types:*.{ts,tsx}": ["@parcel/transformer-typescript-types"]
  }
}
```

Note that you'll need a separate target for the types from the source code for this to work, just like you would have separate "main" and "types" targets in simple library projects.

_Note: I am answering this on mobile and have not tested it. 🙂_

**Comment by captbaritone** [comment]
Thanks for the quick reply. This works. One thing that took me a second to figure out was that I needed to give the target a filename (by creating a top level key in my package.json) that ended with `.d.ts`.

**Comment by Zertz** [comment]
The comments here put me on the right track but I had a poke around quite a bit to get it right. Here's a full example:

**.parcelrc**

```json
{
    "extends": "@parcel/config-default",
    "bundler": "@parcel/bundler-library",
    "transformers": {
        "main-types:*.{ts,tsx}": [
            "@parcel/transformer-typescript-types"
        ]
    }
}
```

**package.json**

```json
{
    "main": "dist/main.js",
    "main-types": "dist/main.d.ts",
    "css": "dist/main.css",
    "targets": {
        "main": {
            "source": "src/main.tsx"
        },
        "main-types": {
            "source": "src/main.tsx"
        },
        "css": {
            "source": "src/main.css"
        }
    }
}
```
